File: consumer-1/tests/mock.test.js
/**
 * MOCK ADAPTER APPROACH
 * ---------------------
 * Generates mock responses based on the OpenAPI schema without making real HTTP calls.
 *
 * Prerequisites:
 * - CVT server running (for schema validation)
 * - No producer service needed
 *
 * When to use:
 * - Unit tests that need to run in isolation
 * - CI pipelines where the producer isn't available
 * - Testing consumer code paths without external dependencies
 */

const { ContractValidator } = require('@sahina/cvt-sdk');
const { createMockAdapter } = require('@sahina/cvt-sdk/adapters');
const path = require('path');

const CVT_SERVER_ADDR = process.env.CVT_SERVER_ADDR || 'localhost:9550';
const SCHEMA_PATH = process.env.SCHEMA_PATH || path.join(__dirname, '../../producer/calculator-api.json');

describe('Mock Client Approach', () => {
  let validator;
  let mock;

  beforeAll(async () => {
    validator = new ContractValidator(CVT_SERVER_ADDR);
    await validator.registerSchema('calculator-api', SCHEMA_PATH);

    mock = createMockAdapter({
      validator,
      cache: true,
      generateOptions: { useExamples: true },
    });
  });

  afterAll(() => {
    if (mock) {
      mock.clearCache();
    }
    if (validator) {
      validator.close();
    }
  });

  beforeEach(() => {
    mock.clearInteractions();
  });

  describe('/add endpoint mocking', () => {
    test('should generate valid mock response for add', async () => {
      const response = await mock.fetch('http://calculator-api/add?x=5&y=3');
      const data = await response.json();

      expect(response.status).toBe(200);
      expect(data).toHaveProperty('result');
      expect(typeof data.result).toBe('number');
    });

    test('should capture mock interactions', async () => {
      await mock.fetch('http://calculator-api/add?x=10&y=20');

      const interactions = mock.getInteractions();
      expect(interactions.length).toBe(1);
      expect(interactions[0].request.method).toBe('GET');
      expect(interactions[0].request.path).toContain('/add');
    });

    test('should generate response matching schema', async () => {
      const response = await mock.fetch('http://calculator-api/add?x=1&y=1');
      const data = await response.json();

      const validationResult = await validator.validate(
        { method: 'GET', path: '/add?x=1&y=1', headers: {} },
        { statusCode: 200, headers: { 'content-type': 'application/json' }, body: data }
      );

      expect(validationResult.valid).toBe(true);
    });
  });

  describe('/subtract endpoint mocking', () => {
    test('should generate valid mock response for subtract', async () => {
      const response = await mock.fetch('http://calculator-api/subtract?x=10&y=4');
      const data = await response.json();

      expect(response.status).toBe(200);
      expect(data).toHaveProperty('result');
      expect(typeof data.result).toBe('number');
    });

    test('should capture subtract interaction', async () => {
      await mock.fetch('http://calculator-api/subtract?x=100&y=50');

      const interactions = mock.getInteractions();
      expect(interactions.length).toBe(1);
      expect(interactions[0].request.path).toContain('/subtract');
    });
  });

  describe('interaction capture for registration', () => {
    test('should capture all Consumer-1 endpoints for registration', async () => {
      await mock.fetch('http://calculator-api/add?x=5&y=3');
      await mock.fetch('http://calculator-api/subtract?x=10&y=4');

      const interactions = mock.getInteractions();
      expect(interactions.length).toBe(2);

      const paths = interactions.map(i => i.request.path);
      expect(paths.some(p => p.includes('/add'))).toBe(true);
      expect(paths.some(p => p.includes('/subtract'))).toBe(true);
    });
  });
});

File: consumer-4/tests/mock_test.go
package tests

import (
	"context"
	"encoding/json"
	"io"
	"testing"

	"github.com/sahina/cvt/sdks/go/cvt"
	"github.com/sahina/cvt/sdks/go/cvt/adapters"
)

// TestMock_AddResponse verifies mock generates valid response for /add
func TestMock_AddResponse(t *testing.T) {
	config := getTestConfig(t)
	validator := newTestValidator(t, config)
	defer validator.Close()

	mock := adapters.NewMock(validator)
	client := mock.Client()

	resp, err := client.Get("http://calculator-api/add")
	if err != nil {
		t.Fatalf("Mock fetch failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		t.Errorf("Expected 200, got %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		t.Fatalf("Failed to read response body: %v", err)
	}
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		t.Fatalf("Failed to parse response: %v", err)
	}
	if _, ok := data["result"]; !ok {
		t.Errorf("Expected 'result' field in response, got: %v", data)
	}
}

// TestMock_SubtractResponse verifies mock generates valid response for /subtract
func TestMock_SubtractResponse(t *testing.T) {
	config := getTestConfig(t)
	validator := newTestValidator(t, config)
	defer validator.Close()

	mock := adapters.NewMock(validator)
	client := mock.Client()

	resp, err := client.Get("http://calculator-api/subtract")
	if err != nil {
		t.Fatalf("Mock fetch failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		t.Errorf("Expected 200, got %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		t.Fatalf("Failed to read response body: %v", err)
	}
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		t.Fatalf("Failed to parse response: %v", err)
	}
	if _, ok := data["result"]; !ok {
		t.Errorf("Expected 'result' field in response, got: %v", data)
	}
}

// TestMock_CapturesInteractions verifies interactions are recorded
func TestMock_CapturesInteractions(t *testing.T) {
	config := getTestConfig(t)
	validator := newTestValidator(t, config)
	defer validator.Close()

	mock := adapters.NewMock(validator)
	mock.ClearInteractions()
	client := mock.Client()

	client.Get("http://calculator-api/add")

	interactions := mock.GetInteractions()
	if len(interactions) != 1 {
		t.Errorf("Expected 1 interaction, got %d", len(interactions))
	}
}

// TestMock_CapturesAllConsumer4Endpoints verifies both add+subtract endpoints are recorded
func TestMock_CapturesAllConsumer4Endpoints(t *testing.T) {
	config := getTestConfig(t)
	validator := newTestValidator(t, config)
	defer validator.Close()

	mock := adapters.NewMock(validator)
	mock.ClearInteractions()
	client := mock.Client()

	client.Get("http://calculator-api/add")
	client.Get("http://calculator-api/subtract")

	interactions := mock.GetInteractions()
	if len(interactions) != 2 {
		t.Errorf("Expected 2 interactions, got %d", len(interactions))
	}
}

// TestMock_ResponseValidatesAgainstSchema verifies mock response matches schema
func TestMock_ResponseValidatesAgainstSchema(t *testing.T) {
	config := getTestConfig(t)
	validator := newTestValidator(t, config)
	defer validator.Close()

	mock := adapters.NewMock(validator)
	client := mock.Client()

	resp, err := client.Get("http://calculator-api/add")
	if err != nil {
		t.Fatalf("Mock fetch failed: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		t.Fatalf("Failed to read response body: %v", err)
	}
	var bodyData map[string]interface{}
	if err := json.Unmarshal(body, &bodyData); err != nil {
		t.Fatalf("Failed to parse response: %v", err)
	}

	ctx := context.Background()
	result, err := validator.Validate(ctx,
		cvt.ValidationRequest{Method: "GET", Path: "/add"},
		cvt.ValidationResponse{StatusCode: 200, Body: bodyData},
	)
	if err != nil {
		t.Fatalf("Validation error: %v", err)
	}
	if !result.Valid {
		t.Errorf("Mock response should validate against schema: %v", result.Errors)
	}
}
